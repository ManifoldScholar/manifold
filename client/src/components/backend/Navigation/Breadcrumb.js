import React, { PureComponent, PropTypes } from 'react';
import { Link } from 'react-router-dom';

export default class Breadcrumb extends PureComponent {

  static displayName = "Navigation.Breadcrumb";

  static propTypes = {
    links: PropTypes.array
  };

  render() {
    return (
      <div className="container flush">
        <nav className="breadcrumb-primary">
          <Link to={this.props.links[0].path} className="initial">
            <i className="manicon manicon-arrow-left"></i>
            {'Back to:'}
          </Link>
          <ul>
            {this.props.links.map((item, index) => {
              return (
                <li key={index}>
                  <Link to={item.path} className="back-link-primary-segment">
                    {item.label}
                  </Link>
                </li>
              );
            })}
          </ul>
        </nav>
      </div>
    );
  }
}
