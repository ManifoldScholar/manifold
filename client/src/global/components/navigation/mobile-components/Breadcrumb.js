import React, { PureComponent } from "react";
import { matchPath } from "react-router-dom";
import { NavLink } from "react-router-dom";
import PropTypes from "prop-types";
import lh from "helpers/linkHandler";
import IconComposer from "global/components/utility/IconComposer";
import { withTranslation } from "react-i18next";

class MobileBreadcrumb extends PureComponent {
  static propTypes = {
    links: PropTypes.array.isRequired,
    location: PropTypes.object.isRequired,
    t: PropTypes.func
  };

  get segments() {
    const journalIsActive = this.props.journalIsActive;
    const segments = [];
    if (typeof journalIsActive !== "boolean") return segments;

    const firstMatch = this.match(this.props.links);
    /* eslint-disable no-nested-ternary */
    const first = journalIsActive
      ? firstMatch.route === "frontendProjects"
        ? { label: "titles.journals", route: "frontendJournals" }
        : firstMatch.route === "backendProjects"
        ? { label: "titles.journals", route: "backendJournals" }
        : firstMatch
      : firstMatch;
    /* eslint-enable no-nested-ternary */

    if (first) {
      segments.push(first);
      const second = this.match(first.children);
      if (second) {
        segments.push(second);
      }
    }
    return segments;
  }

  match(links) {
    if (!links) return null;
    return links.find(link => {
      const route = lh.routeFromName(link.route);

      if (link.matchType === "link" || link.externalUrl) {
        return this.props.location.pathname === this.pathForLink(link);
      }
      return matchPath(this.props.location.pathname, route) !== null;
    });
  }

  pathForLink(link) {
    if (link.externalUrl) return link.externalUrl;
    const args = link.args || [];
    const route = link.linksTo || link.route;
    return lh.link(route, ...args);
  }

  render() {
    let count = 0;
    const segments = this.segments;
    const size = segments.length;

    return (
      <nav className="breadcrumb-list hide-82">
        {this.segments.map(link => {
          count += 1;
          return (
            <span key={count}>
              <NavLink
                className="breadcrumb-list__link"
                to={this.pathForLink(link)}
              >
                {this.props.t(link.label)}
              </NavLink>
              {count < size && (
                <IconComposer
                  icon="disclosureDown16"
                  size="default"
                  className="breadcrumb-list__icon"
                />
              )}
            </span>
          );
        })}
      </nav>
    );
  }
}

export default withTranslation()(MobileBreadcrumb);
