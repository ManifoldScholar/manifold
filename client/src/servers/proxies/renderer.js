import config from "config";
import proxy from "http-proxy-middleware";
import ch from "../../helpers/consoleHelpers";
import createStore from "../../store/createStore";
import exceptionRenderer from "../../helpers/exceptionRenderer";
import manifoldBootstrap from "bootstrap";
import CookieHelper from "helpers/cookie/Server";

const ssrRenderUrl = `http://${config.services.client.domain}:${config.services.client.sparePort}`;

export default function makeRendererProxy(requestHandler) {
  const reactServerProxy = proxy({
    target: ssrRenderUrl,
    changeOrigin: true,
    logLevel: "silent",

    onError: (err, req, res) => {
      ch.error(`Error proxying to port ${ssrRenderUrl}`);
      ch.error(`Perhaps the SSR service is in the process of reloading?`);
      ch.error(`Falling back to SSR rescue service only`);

      const store = createStore();

      const render = () => {
        try {
          requestHandler(req, res);
        } catch (error) {
          if (error.code === "MODULE_NOT_FOUND") {
            const msg =
              "Waiting for initial Webpack build to complete. Wait a few seconds " +
              "and reload.";
            ch.error(msg);
            res.setHeader("Content-Type", "text/html");
            res.end(msg);
          } else {
            ch.error(
              `Rendering fallback failed to render inn server-developmet.js`
            );
            res.setHeader("Content-Type", "text/html");
            res.end(exceptionRenderer(error));
          }
        }
      };

      const promises = [];
      const cookie = new CookieHelper(req, res);
      promises.push(manifoldBootstrap(store.getState, store.dispatch, cookie));
      Promise.all(promises).then(render, render);
    }
  });
  return reactServerProxy;
}
