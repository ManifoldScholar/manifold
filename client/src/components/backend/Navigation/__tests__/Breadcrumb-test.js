import React from 'react';
import renderer from 'react-test-renderer';
import Breadcrumb from '../Breadcrumb';

describe("Navigation.Breadcrumb component", () => {

  const links = [
    {
      path: '/link-1',
      label: 'Link One'
    },
    {
      path: '/link-2',
      label: 'Link Two'
    }
  ]

  it('renders correctly', () => {
    const component = renderer.create(
      <Breadcrumb links={links} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

