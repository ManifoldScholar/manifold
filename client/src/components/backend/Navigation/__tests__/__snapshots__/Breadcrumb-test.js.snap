// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend.Navigation.Breadcrumb component renders correctly 1`] = `
<nav
  className="breadcrumb-primary"
>
  <div
    className="container flush"
  >
    <a
      className="initial"
      href="/link-1"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="manicon manicon-arrow-left"
      />
      Back to:
    </a>
    <ul>
      <li>
        <a
          className="back-link-primary-segment"
          href="/link-1"
          onClick={[Function]}
        >
          Link One
        </a>
      </li>
      <li>
        <a
          className="back-link-primary-segment"
          href="/link-2"
          onClick={[Function]}
        >
          Link Two
        </a>
      </li>
    </ul>
  </div>
</nav>
`;
